### PR TITLE
Add MXFP4 training support via a dequantize-on-the-fly fallback

### LINKED ISSUE #40170: Add MXFP4 MoE/attention backward kernels
## A Call To Action!

The Hugging Face community needs you! 🫵 

The `GPT-OSS` model was recently added using `mxfp4` weights. These 4-bit weights are tiny and highly performant on H100/B100/50xx GPUs, but we only have kernels for **forward** passes. This means that users can't train `GPT-OSS` unless they convert the weights to `bfloat16`. This uses 4X more memory, and reduces speed enormously!

We want **native** `mxfp4` training, but that means we need **backward** kernels too. Ideally, these kernels should also support GPUs that don't have FP4 hardware, so users can still benefit from reduced memory usage during training, even if the computation has to be done in FP8 or `bfloat16`.

### Custom Kernels: An extremely short introduction

`transformers` uses the [Kernels](https://github.com/huggingface/kernels) library to load custom kernels. The critical kernel is the MoE kernel, because attention weights are stored in `bfloat16` for `GPT-OSS`.  This kernel lives on the Hub, in the [triton-kernels repo](https://huggingface.co/kernels-community/triton_kernels). The forward kernel is in the [matmul_ogs file](https://huggingface.co/kernels-community/triton_kernels/blob/main/torch-ext/triton_kernels/matmul_ogs.py), but the backward kernel should probably go in its own file.

This won't be easy - the kernels are written in [Triton](https://triton-lang.org/main/index.html), so this is not a good issue for beginners! This will probably be overwhelming if you don't have some experience with raw CUDA or Triton programming already.

Once the hard work of writing the kernel is done, we can help with integrating it into Transformers (there's a [work-in-progress blogpost](https://github.com/huggingface/blog/pull/2971) about this). We can help with serious attempts, but please don't just throw a code agent at the problem or something (you can use a code agent to help with writing and testing the kernel, but only if you're competent enough to evaluate and bugfix its outputs!)

### PR DESCRIPTION
**Pull request by rexminnis**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47464)

# Add MXFP4 training support via a dequantize-on-the-fly fallback

Fixes #40170 (partially — see "Relation to backward kernels" below), closes the gap behind #40236.

## What this does

Enables LoRA/adapter fine-tuning of MXFP4-quantized models (gpt-oss-20b/120b) **with the quantized weights kept packed**, on hardware where the dequantized-bf16 recipe cannot fit:

- `Mxfp4HfQuantizer.is_trainable` → `True` (with a `warning_once` explaining the fallback + recommending gradient checkpointing).
- `mlp_forward` gains a grad-aware branch: when `torch.is_grad_enabled()` and gradients are required, the MoE block runs the **eager reference math** (top-k router + `GptOssExperts`-style expert loop) over weights **dequantized on the fly**; otherwise the existing triton kernel path runs unchanged. Inference performance is untouched.
- `Mxfp4GptOssExperts` gains `_dequantize_weights()` (layout-safe: `convert_layout(..., StridedLayout)` before `upcast_from_mxfp_torch`, per-expert chunking to bound the fp32 padding intermediates) and `forward_dequantized()` (the autograd-visible expert forward, mirroring the eager `GptOssExperts.forward` math).

The packed MXFP4 storage is never modified. The dequantized tensors are transients owned by the autograd graph; with gradient checkpointing enabled, at most ~one decoder layer's worth (~1.7 GB bf16 for gpt-oss-20b) is alive at a time.

## Why (the silent-gradient trap)

The triton MXFP4 kernels implement forward only. If a user lifts the `is_trainable` gate today (one-line patch, circulating in gists), training **appears to work** — losses go down, grad norms look healthy — but the expert outputs have `grad_fn=None`, so gradients flow through residual connections alone. Every LoRA trained that way learns from biased gradients with no error or warning. We verified this empirically before writing this PR:

```
EXPERT model.layers.0.mlp.experts {'out_grad_fn': None, 'out_requires_grad': False, 'in_requires_grad': True}
VERDICT: DISCONNECTED — residual-only gradients
```

This PR replaces that trap with a correct (if slower) path.

## Validation (gpt-oss-20b, single 24 GB NVIDIA L4, SM89)

The exact code in this PR (injected over a transformers 5.14.1 install) was validated end to end, not a prototype variant:

- **Numerical equivalence**: patched eager-dequant MLP output vs the triton kernel forward on the same input: cosine similarity **0.999992**, relative error **0.4%** (bf16 dequant math vs kernel-internal upcasting).
- **Gradient connectivity**: expert outputs carry `grad_fn`; gradient reaches the embedding output (norm ≈ 2.86) with all base parameters frozen.
- **End-to-end LoRA fine-tune** (axolotl, r=8 on q/k/v/o, seq 256, gradient checkpointing): 5/5 steps, grad_norm 13–44, **peak 14.9 GiB** (13 GB packed weights + transients), ~7.4 s/step at micro-batch 1.
- **Adapter round-trip**: saved adapter (`lora_B` abs-sum 702.7 ≠ 0) loads onto the quantized base via `PeftModel.from_pretrained`, greedy generation visibly shifts vs base; adapter also serves via vLLM (merged into checkpoint, and dynamically on vLLM ≥ 0.25.1).

Hardware note: validated on SM89 where the triton weight layout is already `StridedLayout`; `_dequantize_weights` calls `convert_layout(..., StridedLayout)` explicitly so swizzled layouts (SM90+) should round-trip, but we could not validate on Hopper/Blackwell — reviewer confirmation welcome.

## What this unlocks for users

- gpt-oss-20b fine-tuning on 24 GB cards (4090/L4/Colab-class) instead of 80 GB; gpt-oss-120b (~60 GB packed) plausibly on a single 80 GB device instead of multi-node bf16 (~240 GB).
- Training against the official checkpoint bytes — no third-party requantized copies, no 42 GB dequantized saves; adapters come out vLLM-serving-compatible.
- Axolotl/TRL/PEFT recipes work without custom launchers or monkey-patches.

## Relation to the backward triton kernels (#40170) and to #40180

This is deliberately a **pure-Python fallback**, not the fused MXFP4 backward kernels the issue ultimately wants: ~2–4× slower per step than a fused implementation would be, since the backward recomputes dequantization. It is intended to (a) make training *correct and possible* now, and (b) serve as the numerical reference implementation to validate the real backward kernels against when they land — at which point the grad-aware branch can dispatch to them instead.

To be explicit about scope, in the terms [raised on #40180](https://github.com/huggingface/transformers/pull/40180#discussion_r2282956893) (*"isn't this MXFP4 weight storage and computation in a fallback dtype?"*): **yes — by design, and that split is the contract.** The *storage* half is where the memory win lives and it is fully preserved — the packed weights are never materialized in bf16 (that's the 14.9 GiB peak above, vs ~42 GB for dequantized-bf16 weights alone). The *compute* half is an eager fallback that is **correct**, which is sufficient for the LoRA/adapter use case: the frozen base weights never need `grad_weight` — gradients only need to flow *through* the expert math to the adapters, which the autograd-visible dequant path provides. Native MXFP4 *compute* remains #40170's fused kernels; #40180 was deferred pending those kernels, and this PR intentionally does not presume them — it is the reference they can be validated against.

## Tests

- `test_mxfp4_expert_output_requires_grad` — quantized load, forward under grad; asserts expert output has `grad_fn` and backward reaches inputs (GPU + triton kernels required).
- `test_mxfp4_dequant_forward_matches_kernel` — cosine similarity ≥ 0.999 between the two paths on random inputs.

### CODE REVIEW

**Comment by rexminnis** [comment]
cc @Rocketknight1 @salmanmohammadi — you both reviewed #40180 (native MXFP4 training, deferred pending backward kernels). This PR takes the complementary route: weights **stay packed MXFP4** (15 GB peak for gpt-oss-20b LoRA on a single 24 GB L4, vs 42 GB for bf16 weights alone), and the backward runs eager reference math over per-expert on-the-fly dequantization — so gradients through the expert path are real (cosine 0.999992 vs the Triton forward), avoiding the silent-gradient issue. Scoped to LoRA/adapter training, where the frozen base never needs `grad_weight`; it's intended as the eager reference implementation for validating the eventual #40170 backward kernels rather than a substitute for them. Review would be much appreciated.

**Comment by Rocketknight1** [comment]
Sorry, no code agent PRs from first-time contributors!
